File: backend/src/schemas/script_review.py
"""Strict manual audit records; these inputs never grant publication rights."""
from typing import Annotated, Any, Literal

from pydantic import Field, StringConstraints, model_validator

from src.schemas.script_package import Digest, PackageModel, References, StableId


Category = Literal["PROVENANCE", "TIMELINE", "EVIDENCE", "KNOWLEDGE_BOUNDARY", "PLAYABILITY"]
ReviewText = Annotated[str, StringConstraints(min_length=1, max_length=4000, pattern=r"\S")]


class FindingTarget(PackageModel):
    collection: Literal["introduction", "settlement", "characters", "phases", "knowledge", "evidence", "truth"]
    id: StableId | None

    @model_validator(mode="after")
    def check_identifier(self) -> "FindingTarget":
        if (self.collection in ("introduction", "settlement")) != (self.id is None):
            raise ValueError("invalid finding target")
        return self


class AuditFinding(PackageModel):
    id: StableId
    category: Category
    severity: Literal["BLOCKER", "WARNING", "INFO"]
    target: FindingTarget
    message: ReviewText
    sources: References


class ManualAuditReport(PackageModel):
    schema_version: Literal["script-audit/1.0"]
    summary: ReviewText
    coverage: list[Category] = Field(min_length=1, max_length=5)
    findings: list[AuditFinding] = Field(max_length=200)

    @model_validator(mode="after")
    def unique_records(self) -> "ManualAuditReport":
        if (len(set(self.coverage)) != len(self.coverage)
                or len({item.id for item in self.findings}) != len(self.findings)
                or any(item.category not in self.coverage for item in self.findings)):
            raise ValueError("invalid audit coverage or finding IDs")
        return self


class SubmitAuditRequest(PackageModel):
    idempotency_key: StableId
    expected_package_hash: Digest
    bundle_hash: Digest
    report: ManualAuditReport


class FindingDispositionRequest(PackageModel):
    idempotency_key: StableId
    expected_package_hash: Digest
    expected_audit_hash: Digest
    expected_revision: int = Field(ge=0, le=1000000)
    finding_id: StableId
    status: Literal["OPEN", "ACKNOWLEDGED", "DISMISSED"]
    note: ReviewText


class FindingTargetV12(FindingTarget):
    collection: Literal["introduction", "settlement", "characters", "phases", "knowledge", "evidence", "truth",
                        "mechanics.actions", "mechanics.phase_budgets"]


class AuditFindingV12(AuditFinding):
    target: FindingTargetV12


class ManualAuditReportV12(ManualAuditReport):
    schema_version: Literal["script-audit/1.1"]
    findings: list[AuditFindingV12] = Field(max_length=200)


class SubmitAuditRequestV12(SubmitAuditRequest):
    report: ManualAuditReportV12


class FindingTargetV13(FindingTargetV12):
    collection: Literal["introduction", "settlement", "characters", "phases", "knowledge", "evidence", "truth",
                        "mechanics.actions", "mechanics.phase_budgets", "memories"]


class AuditFindingV13(AuditFindingV12):
    target: FindingTargetV13


class ManualAuditReportV13(ManualAuditReportV12):
    schema_version: Literal["script-audit/1.2"]
    findings: list[AuditFindingV13] = Field(max_length=200)


class SubmitAuditRequestV13(SubmitAuditRequest):
    report: ManualAuditReportV13


def parse_audit_report(document: Any, package_contract: str | None = None) -> ManualAuditReport | ManualAuditReportV12:
    """Keep report versions separate; a 1.2 candidate requires the new report."""
    version = document.get("schema_version") if isinstance(document, dict) else None
    if package_contract is not None:
        expected = {"script-package/1.0": "script-audit/1.0", "script-package/1.1": "script-audit/1.0",
                    "script-package/1.2": "script-audit/1.1", "script-package/1.3": "script-audit/1.2"}.get(package_contract)
        if expected is None or version != expected:
            raise ValueError("audit contract does not match package")
    model = {"script-audit/1.1": ManualAuditReportV12, "script-audit/1.2": ManualAuditReportV13}.get(version, ManualAuditReport)
    return model.model_validate(document)


def parse_submit_audit_request(document: Any) -> SubmitAuditRequest | SubmitAuditRequestV12:
    report = document.get("report") if isinstance(document, dict) else None
    version = report.get("schema_version") if isinstance(report, dict) else None
    model = {"script-audit/1.1": SubmitAuditRequestV12, "script-audit/1.2": SubmitAuditRequestV13}.get(version, SubmitAuditRequest)
    return model.model_validate(document)


def finding_entity(document: dict, target: FindingTarget | FindingTargetV12 | dict) -> dict | None:
    """Resolve only a declared target, with phase_id identifying a budget row."""
    if not isinstance(document, dict):
        return None
    model = {"script-package/1.2": FindingTargetV12, "script-package/1.3": FindingTargetV13}.get(document.get("schema_version"), FindingTarget)
    try:
        parsed = model.model_validate(target.model_dump() if isinstance(target, FindingTarget) else target)
    except (ValueError, TypeError):
        return None
    if parsed.collection == "introduction":
        result = document.get("introduction")
    elif parsed.collection == "settlement":
        settlement = document.get("settlement")
        result = settlement.get("instructions") if isinstance(settlement, dict) else None
    else:
        key = "id"
        if parsed.collection.startswith("mechanics."):
            mechanics = document.get("mechanics")
            collection = parsed.collection.split(".")[1]
            rows = mechanics.get(collection) if isinstance(mechanics, dict) else None
            if collection == "phase_budgets":
                key = "phase_id"
        else:
            rows = document.get(parsed.collection)
        if not isinstance(rows, list):
            return None
        matches = [item for item in rows if isinstance(item, dict) and item.get(key) == parsed.id]
        result = matches[0] if len(matches) == 1 else None
    return result if isinstance(result, dict) else None

File: frontend/src/types/scriptReview.ts
export type ReviewCategory = 'PROVENANCE' | 'TIMELINE' | 'EVIDENCE' | 'KNOWLEDGE_BOUNDARY' | 'PLAYABILITY';
export type FindingStatus = 'OPEN' | 'ACKNOWLEDGED' | 'DISMISSED';

export type ReviewCandidate = {
  id: number;
  script_key: string;
  content_version: string;
  title: string;
  package_hash: string;
  contract_version: string;
};

export type ReviewFinding = {
  id: string;
  category: ReviewCategory;
  severity: 'BLOCKER' | 'WARNING' | 'INFO';
  target: {
    collection: 'introduction' | 'settlement' | 'characters' | 'phases' | 'knowledge' | 'evidence' | 'truth'
      | 'mechanics.actions' | 'mechanics.phase_budgets' | 'memories';
    id: string | null;
  };
  message: string;
  sources: { source_id: string; page?: number | null; anchor?: string | null }[];
};

export type ScriptAuditReport = {
  schema_version: 'script-audit/1.0' | 'script-audit/1.1' | 'script-audit/1.2';
  summary: string;
  coverage: ReviewCategory[];
  findings: ReviewFinding[];
};

export type FindingDisposition = {
  id: number;
  revision: number;
  finding_id: string;
  status: FindingStatus;
  note: string;
  submitted_by: number;
};

export type ScriptAudit = {
  id: number;
  audit_hash: string;
  version_id: number;
  package_hash: string;
  bundle_hash: string;
  source_report_hash: string;
  submitted_by: number;
  report: ScriptAuditReport;
  revision: number;
  dispositions: FindingDisposition[];
  open_blockers: number;
  open_warnings: number;
  publication_ready: false;
};

export type CandidateReview = {
  candidate: ReviewCandidate;
  audits: ScriptAudit[];
  references: { target: ReviewFinding['target']; sources: ReviewFinding['sources'] }[];
  source_files: { id: string; relative_path: string; kind: 'original' | 'normalized' | 'supplement' }[];
};
export type FindingDraft = { status: FindingStatus; note: string };

export type RuleReview = {
  schema_version: 'rule-review/1.0' | 'rule-review/1.1';
  runtime_facts?: { package_contract:'script-package/1.2'; human_players:1; investigation_actor:'SELECTED_HUMAN_ONLY';
    action_success_limit:'ONCE_PER_SESSION'; phase_budget:'SHARED_NO_CARRY'; material_recipient:'DECLARED_OWNER' };
  package_hash: string;
  bundle_hash: string;
  semantic_status: 'UNREVIEWED';
  publication_ready: false;
  row_count: number;
  offset: number;
  limit: number;
  next_offset: number | null;
  rows: {
    target: { collection: string; id: string | null };
    label: string;
    rules: Record<string, unknown>;
    source_count: number;
    sources_truncated: boolean;
    sources: { source_id: string; kind: string; anchor: string | null; page: number | null;
      text: string | null; status: string; truncated: boolean }[];
    notices: { code: string; action_id: string; evidence_id: string }[];
    notice_count: number;
    notices_truncated: boolean;
  }[];
};

export type SubmitAuditRequest = {
  idempotency_key: string;
  expected_package_hash: string;
  bundle_hash: string;
  report: ScriptAuditReport;
};

export type SubmitDispositionRequest = {
  idempotency_key: string;
  expected_package_hash: string;
  expected_audit_hash: string;
  expected_revision: number;
  finding_id: string;
  status: FindingStatus;
  note: string;
};
